### PR TITLE
Specialize Adaptive's allocator Recycler based on magazine's owner

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:
EventLoop local magazines are treated the same as shared ones

Modifications:
Use existing Recycler(s) on un-shared magazines to save expensive atomic operations while reusing adaptive buffers

Result:
Further improve https://github.com/netty/netty/issues/14279 for the common path i.e. event loops

### CODE REVIEW

**Comment by franz1981** [comment]
@normanmaurer this is on top of #14414 just to ease reviewing it

**Comment by franz1981** [comment]
@normanmaurer @chrisvest I've gotten rid of the megamorphic call mentioned at https://github.com/netty/netty/pull/14421#issuecomment-2441531926 by adding 2c3b15c50daed0f4b6a93a34568e2e9f6c304520

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
I tried running benchmarks with this, but I'm not seeing much improvement. I used 10 forks in these runs:

#14414:

```
Benchmark                                             (size)   Mode  Cnt         Score        Error  Units
ByteBufAllocatorBenchmark.adaptiveDirectAllocAndFree   00256  thrpt  100  18495157.592 ± 253756.522  ops/s
ByteBufAllocatorBenchmark.adaptiveHeapAllocAndFree     00256  thrpt  100  20277970.413 ± 428136.189  ops/s

Benchmark                                                    (size)   Mode  Cnt         Score         Error  Units
ByteBufAllocatorConcurrentBenchmark.allocateReleaseAdaptive   00256  thrpt  100  79617949.492 ± 6718078.227  ops/s
```

#14421:

```
Benchmark                                             (size)   Mode  Cnt         Score         Error  Units
ByteBufAllocatorBenchmark.adaptiveDirectAllocAndFree   00256  thrpt  100  16641428.737 ± 1887660.133  ops/s
ByteBufAllocatorBenchmark.adaptiveHeapAllocAndFree     00256  thrpt  100  20258834.173 ±  112095.227  ops/s

Benchmark                                                    (size)   Mode  Cnt         Score         Error  Units
ByteBufAllocatorConcurrentBenchmark.allocateReleaseAdaptive   00256  thrpt  100  79411577.078 ± 9787056.407  ops/s
```

**Comment by franz1981** [comment]
Are u using an apple processor? 
They way apple handle u contended atomic is slightly different from x86; I will re run (with profiling) the same benchmark, in case.
The other reason is that something in fast thread local doesn't make it as good as we hope for this use case OR the benchmark is not stressing what we believe it stress.

**Comment by franz1981** [comment]
As expected by my own comment at

> OR the benchmark is not stressing what we believe it stress (this one seems the more likely to me)

 running the `ByteBufAllocatorConcurrentBenchmar` benchmark on this branch shows that something is off:

image

which means that despite the jmh executor should be an event loop, the magazine used is NOT event loop local.

**Comment by franz1981** [comment]
And it seems @chrisvest that you need to set 

and the actual behaviour reflect how powerful this optimization is

image

As you can see the number of atomic ops are much reduced and numbers have changed accordly

This PR without 
```
Benchmark                                                    (size)   Mode  Cnt        Score        Error  Units
ByteBufAllocatorConcurrentBenchmark.allocateReleaseAdaptive   00256  thrpt   20  8959741.229 ± 149752.335  ops/s
```
Now, adding the proper configuration i.e. `-Dio.netty.allocator.useCachedMagazinesForNonEventLoopThreads=true`
in order to make the custom harness to be considered OK despite not been an event loop.

This PR:
```
Benchmark                                                    (size)   Mode  Cnt         Score        Error  Units
ByteBufAllocatorConcurrentBenchmark.allocateReleaseAdaptive   00256  thrpt   20  13629020.463 ± 801453.039  ops/s
```
Before this PR:
```
Benchmark                                                    (size)   Mode  Cnt         Score        Error  Units
ByteBufAllocatorConcurrentBenchmark.allocateReleaseAdaptive   00256  thrpt   20  11208779.418 ± 210872.723  ops/s
```
There's some improvement (as expected) - but I still see too many `compareAndSet` in the profiling data shown in this comment; investigating.

**Comment by franz1981** [comment]
As per 

> There's some improvement (as expected) - but I still see too many compareAndSet in the profiling data shown in this comment; investigating.

I was wrong: it seems that we have 2 compareAndSet (which are too much IMO) - one for `Chunk::release` and one (which "shield" it) on the adaptive `ByteBuf` itself (on `release`) :"(
I would like to have just one and this will further "fly"

**Review by normanmaurer — APPROVED** [review]

**Comment by normanmaurer** [comment]
Should we merge https://github.com/netty/netty/pull/14414 first and just rebase this one as a "follow up" ?

**Comment by franz1981** [comment]
sure things, I would wait @chrisvest feedbacks after adding that option I suggested; x86 is not the only arch we would like to make happy u think 🤔

**Comment by chrisvest** [comment]
> Are u using an apple processor?

Yes.

I reran the benchmark with the option, and see some improvements as well:

```
#14414
Benchmark                                                    (size)   Mode  Cnt         Score         Error  Units
ByteBufAllocatorConcurrentBenchmark.allocateReleaseAdaptive   00256  thrpt  100  83700665.953 ± 9595346.585  ops/s

#14421
Benchmark                                                    (size)   Mode  Cnt         Score          Error  Units
ByteBufAllocatorConcurrentBenchmark.allocateReleaseAdaptive   00256  thrpt  100  95217417.876 ± 12041757.914  ops/s
```

We need to look into why the JMH executor setting isn't having the effect we expect.

**Comment by normanmaurer** [comment]
Please do @chrisvest

**Review by normanmaurer — APPROVED** [review]
LGTM

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@franz1981 feel free to pull in

File: buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java [hunk 1]
```diff
@@ -677,13 +688,17 @@ void free() {
...
+                buf = EVENT_LOOP_LOCAL_BUFFER_POOL.get();
             } else {
-                buf.resetRefCnt();
-                buf.discardMarks();
+                buf = bufferQueue.poll();
+                if (buf == null) {
+                    buf = new AdaptiveByteBuf(handle);
+                }
             }
+            buf.resetRefCnt();
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:700]: I guess its not a big deal but we could just not do this when we allocate a new buffer...

Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:700]: more of the times we end-up reusing some (hopefully) - but I'm ok to move it in the right branch too, LMK

Comment (franz1981) [reply buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java:700]: Anyway is not using a volatile operation, but a lazySet AFAIK, which is way cheaper

**Pull request merged**